### PR TITLE
Some minor LVFormatter::measureText() changes

### PR DESCRIPTION
**Pull request by bbshelper**
I'm finalizing a major optimization to `LVFormatter::measureText()`, so I'd like these smaller ones to be reviewed and merged first.

The first 2 commits have around 0.5% and 0.2% performance gain in my Load and Render test.

### CODE REVIEW

File: crengine/src/lvtextfm.cpp [hunk 1]
```diff
@@ -2422,14 +2406,17 @@ class LVFormatter {
...
-                int dx = tabPosition - m_widths[tabIndex];
-                for ( i=tabIndex; i<m_length; i++ )
-                    m_widths[i] += dx;
+            for (int j = 0; j < m_length && m_widths[j] >= tabPosition; ++j) {
+                if (m_text[j] == '\t') {
+                    int dx = tabPosition - m_widths[j];
+                    for ( int i=j; i<m_length; i++ )
+                        m_widths[i] += dx;
+                    break;
+                }
```
Comment (poire-z) [inline crengine/src/lvtextfm.cpp:2419]: Not really familiar with what that does (even now just trying to understand the purpose).
I get that now, we won't do anything as long as `m_srcs[0]->indent < 0` is false, and if I trust my comment, this should rarely be true. So ok with avoiding these checks in the main loop.
I also think that you too probably don't get what this does, and that you just tried to ensure the same logic (looks like you did), so trusting you :)

In crengine, there are mostly only `i++`, `--i` is rare. As in the inner loop you kept the `i++`, may be make the outer loop also use the `j++` style.

File: crengine/src/lvtextfm.cpp [hunk 2]
```diff
@@ -1976,11 +1958,13 @@ class LVFormatter {
...
-                        if ( lastDirection < 0 )
-                            hints |= LFNT_HINT_DIRECTION_IS_RTL;
-                    }
+                    #if (USE_FRIBIDI==1)
+                        if (m_has_bidi) {
+                            hints |= LFNT_HINT_DIRECTION_KNOWN;
+                            if (FRIBIDI_LEVEL_IS_RTL(lastBidiLevel))
+                                hints |= LFNT_HINT_DIRECTION_IS_RTL;
+                        }
+                    #endif
```
Comment (poire-z) [inline crengine/src/lvtextfm.cpp:1967]: It looks like in the existing code, we set `hints |= LFNT_HINT_DIRECTION_KNOWN` always (when ` #if (USE_FRIBIDI==1)`, so: always).
Now, you set it only when `if (m_has_bidi)`. I guess this first `hints |=` should be moved above.

Ok, I get what you're doing. The commit message should be:
Calculate bidi direction on ~~measuring~~ segment change only, not on every char.

The other (now it is a second call) `FRIBIDI_LEVEL_IS_RTL(lastBidiLevel)` below would only be done when PAD_CHAR_INDEX (inline margin/border/padding) which should be rare.

(Is that FRIBIDI_LEVEL_IS_RTL() really expensive, and worth the pain for you and me ? :) It's just `#define FRIBIDI_LEVEL_IS_RTL(lev) ((lev) & 1)`).

File: crengine/src/lvtextfm.cpp [hunk 3]
```diff
@@ -1845,14 +1845,12 @@ class LVFormatter {
...
         static lUInt8 flags[MAX_TEXT_CHUNK_SIZE+1];
-        int tabIndex = -1;
         #if (USE_FRIBIDI==1)
             FriBidiLevel lastBidiLevel = 0;
             FriBidiLevel newBidiLevel;
         #endif
         #if (USE_HARFBUZZ==1)
-            bool checkIfHarfbuzz = true;
-            bool usingHarfbuzz = false;
+            bool usingHarfbuzz = m_kerning_mode == KERNING_MODE_HARFBUZZ;
```
Comment (poire-z) [inline crengine/src/lvtextfm.cpp:1853]:
```
It doesn't seem to be necessary to set usingHarfbuzz on the first font
met: the variable doesn't change, and when it's checked, there is always
a valid font.
```
Ok, right.
These `usingHarfbuzz` were introcuded by 737f37e2d in 2019.
`m_kerning_mode` by a7cea02c0 only in 2022. I could have removed all that at the time, so better late than never :)

**Review by poire-z — COMMENTED** [review]
I'm not sure these microoptimizations are worth your time and the 60 minutes I had to spend on reviewing to get back into the code and made sure you did not change the behaviour :/

> I'm finalizing a major optimization to LVFormatter::measureText()

I'm really fearing that :\ (If you PR that and don't hear from me, it's because I'll soon be on vacation, not because of fear :))

**Comment by bbshelper** [comment]
> I'm not sure these microoptimizations are worth your time and the 60 minutes I had to spend on reviewing to get back into the code

I'm not sure either :) I must have spent more time optimizing crengine than what I can ever save in reading on e-ink in my lifetime.

> and made sure you did not change the behaviour :/

I try very hard not to. I've setup a  regression test, which is quite useful in catching corner cases. That said, I don't really know a thing about bidi and harfbuzz, and they are not covered by my regression test.

> I'm really fearing that :\ (If you PR that and don't hear from me, it's because I'll soon be on vacation, not because of fear :))

I promise that the new code is clearer than the current one. It's a major _optimization_ and not a rewrite. :)

File: crengine/src/lvtextfm.cpp (hunk 2 above)
Comment (bbshelper) [reply crengine/src/lvtextfm.cpp:1967]: The intention was to remove writes to `lastDirection` in the main loop. As you've pointed out, this commit is problematic. I'll rework it.

File: crengine/src/lvtextfm.cpp (hunk 3 above)
Comment (bbshelper) [reply crengine/src/lvtextfm.cpp:1853]: glad to hear that :)
